File: python/mirage/commands/builtin/generic/uniq.py
from collections.abc import AsyncIterator, Callable

from mirage.commands.builtin.utils.stream import _resolve_source
from mirage.io.async_line_iterator import AsyncLineIterator
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _comparison_key(
    line: bytes,
    skip_fields: int,
    skip_chars: int,
    check_chars: int,
    ignore_case: bool,
) -> bytes:
    text = line
    if skip_fields > 0:
        decoded = text.decode(errors="replace")
        parts = decoded.split()
        remaining = parts[skip_fields:] if skip_fields < len(parts) else []
        text = " ".join(remaining).encode()
    if skip_chars > 0:
        text = text[skip_chars:]
    if check_chars > 0:
        text = text[:check_chars]
    if ignore_case:
        text = text.lower()
    return text


def _should_emit(prev_count: int, duplicates_only: bool,
                 unique_only: bool) -> bool:
    if duplicates_only and prev_count == 1:
        return False
    if unique_only and prev_count > 1:
        return False
    return True


async def _uniq_stream(
    source: AsyncIterator[bytes],
    count: bool,
    duplicates_only: bool,
    unique_only: bool,
    skip_fields: int,
    skip_chars: int,
    ignore_case: bool,
    check_chars: int,
) -> AsyncIterator[bytes]:
    prev_line: bytes | None = None
    prev_key: bytes | None = None
    prev_count = 0
    async for raw_line in AsyncLineIterator(source):
        key = _comparison_key(raw_line, skip_fields, skip_chars, check_chars,
                              ignore_case)
        if key == prev_key:
            prev_count += 1
        else:
            if prev_line is not None and _should_emit(
                    prev_count, duplicates_only, unique_only):
                if count:
                    yield f"{prev_count:>7} ".encode() + prev_line + b"\n"
                else:
                    yield prev_line + b"\n"
            prev_line = raw_line
            prev_key = key
            prev_count = 1
    if prev_line is not None and _should_emit(prev_count, duplicates_only,
                                              unique_only):
        if count:
            yield f"{prev_count:>7} ".encode() + prev_line + b"\n"
        else:
            yield prev_line + b"\n"


async def uniq(
    paths: list[PathSpec],
    *,
    read_stream: Callable[..., AsyncIterator[bytes]],
    accessor: object = None,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    count: bool = False,
    duplicates_only: bool = False,
    unique_only: bool = False,
    skip_fields: int = 0,
    skip_chars: int = 0,
    ignore_case: bool = False,
    check_chars: int = 0,
) -> tuple[ByteSource | None, IOResult]:
    cache: list[str] = []
    if paths:
        source = read_stream(accessor, paths[0])
        cache = [paths[0].strip_prefix]
    else:
        source = _resolve_source(stdin, "uniq: missing operand")

    return _uniq_stream(
        source,
        count=count,
        duplicates_only=duplicates_only,
        unique_only=unique_only,
        skip_fields=skip_fields,
        skip_chars=skip_chars,
        ignore_case=ignore_case,
        check_chars=check_chars,
    ), IOResult(cache=cache)


__all__ = ["uniq"]

File: typescript/packages/core/src/commands/builtin/generic/uniq.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { AsyncLineIterator } from '../../../io/async_line_iterator.ts'
import { IOResult } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'
import { resolveSource } from '../utils/stream.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

interface UniqOptions {
  count: boolean
  duplicatesOnly: boolean
  uniqueOnly: boolean
  skipFields: number
  skipChars: number
  checkChars: number
  ignoreCase: boolean
}

function comparisonKey(line: Uint8Array, opts: UniqOptions): string {
  let text = DEC.decode(line)
  if (opts.skipFields > 0) {
    const parts = text.split(/\s+/).filter((s) => s !== '')
    const remaining = opts.skipFields < parts.length ? parts.slice(opts.skipFields) : []
    text = remaining.join(' ')
  }
  if (opts.skipChars > 0) text = text.slice(opts.skipChars)
  if (opts.checkChars > 0) text = text.slice(0, opts.checkChars)
  if (opts.ignoreCase) text = text.toLowerCase()
  return text
}

function padLeft(value: string, width: number): string {
  return value.length >= width ? value : ' '.repeat(width - value.length) + value
}

function emitLine(line: Uint8Array, count: number, opts: UniqOptions): Uint8Array | null {
  if (opts.duplicatesOnly && count === 1) return null
  if (opts.uniqueOnly && count > 1) return null
  if (opts.count) {
    const prefix = ENC.encode(`${padLeft(String(count), 7)} `)
    const out = new Uint8Array(prefix.byteLength + line.byteLength + 1)
    out.set(prefix, 0)
    out.set(line, prefix.byteLength)
    out[out.byteLength - 1] = 0x0a
    return out
  }
  const out = new Uint8Array(line.byteLength + 1)
  out.set(line, 0)
  out[line.byteLength] = 0x0a
  return out
}

async function* uniqStream(
  source: AsyncIterable<Uint8Array>,
  opts: UniqOptions,
): AsyncIterable<Uint8Array> {
  let prevLine: Uint8Array | null = null
  let prevKey: string | null = null
  let prevCount = 0
  const iter = new AsyncLineIterator(source)
  for await (const rawLine of iter) {
    const key = comparisonKey(rawLine, opts)
    if (key === prevKey) {
      prevCount += 1
    } else {
      if (prevLine !== null) {
        const chunk = emitLine(prevLine, prevCount, opts)
        if (chunk !== null) yield chunk
      }
      prevLine = rawLine
      prevKey = key
      prevCount = 1
    }
  }
  if (prevLine !== null) {
    const chunk = emitLine(prevLine, prevCount, opts)
    if (chunk !== null) yield chunk
  }
}

function parseOptions(flags: Record<string, string | boolean>): UniqOptions {
  const intFlag = (key: 'f' | 's' | 'w'): number =>
    typeof flags[key] === 'string' ? Number.parseInt(flags[key], 10) : 0
  return {
    count: flags.c === true,
    duplicatesOnly: flags.d === true,
    uniqueOnly: flags.u === true,
    skipFields: intFlag('f'),
    skipChars: intFlag('s'),
    checkChars: intFlag('w'),
    ignoreCase: flags.i === true,
  }
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function uniqGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
): Promise<CommandFnResult> {
  const uniqOpts = parseOptions(opts.flags)
  if (paths.length > 0) {
    const first = paths[0]
    if (first === undefined) return [null, new IOResult()]
    return [uniqStream(stream(first), uniqOpts), new IOResult({ cache: [first.stripPrefix] })]
  }
  try {
    const source = resolveSource(opts.stdin, 'uniq: missing operand')
    return [uniqStream(source, uniqOpts), new IOResult()]
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
  }
}
